segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -0,0 +1,263 @@
+        if (count == 0) {
+            log.debug("{} deleteGarbage - no work - sleeping for {}.", chunkedSegmentStorage.getLogPrefix(), config.getGarbageCollectionDelay());
+            return Futures.delayedFuture(config.getGarbageCollectionSleep(), executor);
+        }
+
+        // For each chunk delete if the chunk is not present at all in the metadata or is present but marked as inactive.
+        ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
+        for (val infoToDelete : chunksToDelete) {
+            val chunkToDelete = infoToDelete.name;
+            val txn = chunkedSegmentStorage.getMetadataStore().beginTransaction(false, chunkToDelete);
+            val future =
+                    txn.get(infoToDelete.name)
+                    .thenComposeAsync( metadata -> {
+                            val chunkMetadata = (ChunkMetadata) metadata;
+                            // Delete if the chunk is not present at all in the metadata or is present but marked as inactive.
+                            boolean shouldDelete = (null == chunkMetadata) ? true : !chunkMetadata.isActive();
+                            // Check whether the chunk is marked as inactive
+                            if (shouldDelete) {
+                                return chunkedSegmentStorage.getChunkStorage().openWrite(chunkToDelete)
+                                        .thenComposeAsync(chunkedSegmentStorage.getChunkStorage()::delete, executor)
+                                        .thenRunAsync(() -> {
+                                            if (null != metadata) {
+                                                txn.delete(chunkToDelete);
+                                            }
+                                            log.debug("{} deleteGarbage - deleted chunk={}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                        }, executor)
+                                        .thenComposeAsync(v -> txn.commit(), executor)
+                                        .handleAsync((v, e) -> {
+                                            if (e != null) {
+                                                val ex = Exceptions.unwrap(e);
+                                                if (ex instanceof ChunkNotFoundException) {
+                                                    // Ignore - nothing to do here.
+                                                    log.debug("{} deleteGarbage - Could not delete garbage chunk {}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                                } else {
+                                                    log.warn("{} deleteGarbage - Could not delete garbage chunk {}.", chunkedSegmentStorage.getLogPrefix(), chunkToDelete);
+                                                    // Queue it back.
+                                                    addToGarbage(Collections.singleton(chunkToDelete));
```

Comment:
If we have, let's say, an `IOError` on that chunk that is persistent, does it mean that we will try to garbage collect it forever?

Reply:
yes.

Reply:
If attempt to delete failed then we should retry.